### ISSUE TITLE
unexpected updates in error management / master 2019-08-03

### ISSUE THREAD

**Issue by petrovr** [openssl/openssl#9522]
After synchronization with master I note some issues with recent updates.

1.
openssl/err.h defines ERR_PUT_func_error as ERR_put_func_error
So far so good but was is 
util/libcrypto.num:ERR_put_func_error                      4799 3_0_0   NOEXIST::FUNCTION:
There is nothing in 1.1.1 stable branch.

2.
openssl/err.h defines ERR_PUT_error as ERR_put_error.
ERR_put_error is deprecated . Why ERR_PUT_error is not deprecated as well?

3.
ERR_put_error uses file and line only if build is without no-err 
```
# ifndef OPENSSL_NO_ERR
#  define ERR_PUT_error(l,f,r,fn,ln)      ERR_put_error(l,f,r,fn,ln)
...
# else
#  define ERR_PUT_error(l,f,r,fn,ln)      ERR_put_error(l,f,r,NULL,0)
...
# endif
```

There is no similar functionality with new methods, actually defines **ERR_raise** and especially **ERR_raise_data**.
With other words why ERR_set_debug in ERR_raise_data is called unconditionally?

4.
Related to 3. Now ERR_put_error is macro and again macro also call unconditionally ERR_set_debug. As is visible in 3. file and line are not used in some cases.

**Comment by levitte** [comment]
1. ERR_put_func_error only existed in master for a short while, never in any release. ERR_raise and ERR_raise_data supercede it.
2. Good catch. It should be deprecated as well. Will happen in a future PR 
3. I will have to re-check, but I believe that the definition of OPENSSL_FILE and OPENSSL_LINE is conditional, which achieves the same thing. 
4. Same answer. 

I'm away from keyboard until Thursday, so can't do much about this until then. However, PR welcome if anyone feels inclined.

**Comment by petrovr** [comment]
Richard Levitte wrote:
> 1. ERR_put_func_error only existed in master for a short while, never in any release. ERR_raise and ERR_raise_data supercede it.
Ok. Look like added after my last "synchronization" (~ monthly basis).


> 2. Good catch. It should be deprecated as well. Will happen in a future PR

Solution depend from other points.


> 3. I will have to re-check, but I believe that the definition of OPENSSL_FILE and OPENSSL_LINE is conditional, which achieves the same thing.

no-err vs no-filenames?

So if new model is to split no-err into no-err (no lot load error 
strings) and no-filenames (do not print file in error) solution for 2 
could be different.

a) Now it seems to me ERR_PUT_error could be rewritten to use 
unconditionally ERR_raise.
Less impact over third party code. Note that before FOOerr was 
implemented with use ERR_PUT_error.

b) or may be deprecation is better, i.e. to force engines and etc. to 
use only new functionality.
Is such case no-err condition could be replaced with new rule as file 
and line are macro parameters.


> 4. Same answer.

Due to 3. there is no need to keep ERR_put_error as macro.


As now file and line are subject of build parameter I think that 
ERR_set_debug is responsible to output or not those values. If third 
party code call function with those arguments they should be ignored.



> 
> I'm away from keyboard until Thursday, so can't do much about this until then. However, PR welcome if anyone feels inclined.

There is nothing critical in this issue.
Let first design.


Regards,
Roumen Petrov

**Comment by petrovr** [comment]
Any new on this?
Note that there is two issue in the same area.

**Comment by levitte** [comment]
Thanks for the reminder.  Fixes in #9747

**Comment by petrovr** [comment]
Hmm , the code that sets ERR_DBG_* macros does not look good to me.
It is based on the fact that third party software will use only provided macros. Functions are public and programs could call them directly.
Dunno why  RT3628 manage *_FILE/*_LINE as defines instead to remove code that hold them.

Note doubling -  #9530  .

For *_FUNC I prefer  to discuss in new issue.
